### PR TITLE
Add FFT based on cuFFT

### PR DESCRIPTION
**Pull request by bmerry**
This is based on the FFT implementation in katsdpimager, but using ctypes to interface with CUFFT directly, rather than using scikit-cuda (which has some issues). The unit tests have also been improved.

### CODE REVIEW

File: katsdpsigproc/fft.py [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+                int(np.product(padded_shape_src[-N:])),   # idist
+                arr_type(*padded_shape_dest[-N:]),        # onembed
+                1,                                        # ostride
+                int(np.product(padded_shape_dest[-N:])),  # odist
+                fft_type,                                 # type
+                batches,                                  # batch
+                ctypes.byref(work_size)                   # workSize
+            )
+        self._work_size = work_size.value
+        self._cufft = cufft
```
[the remark this answers is not in the document]
Comment (bmerry) [reply katsdpsigproc/fft.py:296]: I think I just put all the code that's assigning internally-computed values to object attributes together. I'm happy to move it if you think that'll be better.

File: katsdpsigproc/fft.py [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+        elif dtype_src == np.complex128 and dtype_dest == np.complex128:
+            fft_type = _Cufft.cufftType.CUFFT_Z2Z
+        else:
+            raise ValueError("Invalid combination of dtypes")
+
+        cufft = _Cufft()
+        self.context: cuda.Context = context
+        self.shape = shape
+        self.dtype_src = np.dtype(dtype_src)
+        self.dtype_dest = np.dtype(dtype_dest)
```
[the remark this answers is not in the document]
Comment (bmerry) [reply katsdpsigproc/fft.py:262]: I'll look into it - I didn't get any problems with pre-commit but Jenkins is probably not running exactly the same versions.

File: katsdpsigproc/fft.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (james-smith-za) [reply katsdpsigproc/fft.py:296]: "Better" is not an objective term.

I can see pros and cons, but nothing really momentous either way.

I don't think you need to change it, I just wanted to understand your reasoning.

**Review by james-smith-za — APPROVED** [review]
It's at your discretion to fix the mypy issue, I don't think there's anything preventing the code from being merged, but you also don't want to get in the habit of ignoring the fact that the checks have failed, IMO.

**Comment by bmerry** [comment]
> It's at your discretion to fix the mypy issue, I don't think there's anything preventing the code from being merged, but you also don't want to get in the habit of ignoring the fact that the checks have failed, IMO.

Oh, no chance I'll merge without getting it fixed.

File: katsdpsigproc/fft.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (bmerry) [reply katsdpsigproc/fft.py]: It was caused by Jenkins using an older numpy. I didn't investigate exactly why the older numpy needs the annotations, but 1.20 was the first version with annotations and there were a few wrinkles.

**Review by james-smith-za — APPROVED** [review]

**Comment by bmerry** [comment]
> Perhaps it's naive to suggest this, but would it not be better to bump up the version Jenkins is using rather than rolling pre-commit back?

The Jenkins version is shared across all the SDP projects, so bumping it has a lot of potential to break other projects (and to cause version mismatches in other projects that have matched to Jenkins). I did want to eventually change Jenkins to just run pre-commit, but probably won't get around to it.

**Pull request merged**